lemur/plugins/lemur_gcp/certificates.py

```diff
@@ -104,3 +104,29 @@ def parse_certificate_meta(certificate_meta):
         chain="\n".join(chain[1:]),
         name=certificate_meta.name,
     )
+
+
+def get_self_link(project, name):
+    return f"https://www.googleapis.com/compute/v1/projects/{project}/global/sslCertificates/{name}"
+
+
+def calc_diff(certs, new_cert, old_cert):
+    """
+    Produces a list of certificate self-links where new_cert is added and old_cert is removed, if it exists.
+    :param certs:
+    :param new_cert:
+    :param old_cert:
+    :return:
+    """
+    # Shallow copy the list of self-links (strings)
+    result = list(certs)
+    new_cert_idx = -1
+    for idx, self_link in enumerate(certs):
+        if self_link == old_cert:
+            new_cert_idx = idx
+            break
+    if new_cert_idx != -1:
+        result[new_cert_idx] = new_cert
+    else:
+        result.append(new_cert)
```

Comment:
Are there any circumstances where it expected for the old certificate being replaced to not exist in the list? We might want to raise an exception here instead.

Reply:
If the old certificate being replaced does not exist at all in Lemur, the error is surfaced before it hits the rotate logic.

If the old certificate being replaced does not exist on the endpoint based on what's in the DB, we set the new cert to be the default (so it will not use `calc_diff`)

If the old certificate being replaced does exist on the endpoint based on what's in the DB but it does not exist on the LB (deleted out-of-band directly in cloud provider before another source-sync happens), I think it should be a no-op instead of an exception. The AWS plugin's cert rotation is idempotent and ignores removing a cert that isn't already attached.

Reply:
Maybe add a warning log in this case then to indicate that the current DB state is possibly out of sync, but Lemur performed the rotation operations anyway because it was deemed safe to do so.